### PR TITLE
Button timer

### PR DESCRIPTION
**Pull request by lhbrennan**
#### Description

Build new `Button Timed` component: https://zeroheight.com/6d2425e9f/p/81070d-button-timed/b/284ec6

#### Scope
Minor: New Feature

### CODE REVIEW

File: src/button-timed/button-timed.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    interval = setInterval(() => {
+      if (secondsRemaining > 0) {
+        setSecondsRemaining((seconds) => seconds - 1);
+      }
+    }, 1000);
+    if (secondsRemaining === 0) {
+      onClickProp();
+    }
+    return () => clearInterval(interval);
+  }, [secondsRemaining]);
```
Comment (chasestarr) [inline src/button-timed/button-timed.tsx:46]: Should this start counting immediately or should it be initialized somehow?

**Review by chasestarr — COMMENTED** [review]
The visuals look good to me. I'm most interested in how a developer should be using this component. Currently it seems to be set up where when loaded to the page it will begin counting down from the set time and firing an event when finished. I'd like to see some way to stop/start the timer from outside the component.

An alternative api could be something similar to progress-bar where the application controls the progress through the timer and visuals are a pure function of props. Would that be too burdensome?

**Comment by lhbrennan** [comment]
> An alternative api could be something similar to progress-bar where the application controls the progress through the timer and visuals are a pure function of props. Would that be too burdensome?

Yea that would be more flexible. In terms of the API, I think we'd want something like `timeElapsed` (or maybe `currentTime`), as well as something like `totalTime` (or `targetTime`). Thoughts on naming those props? Should we use "seconds" instead of "time" to be more explicit?

**Comment by chasestarr** [comment]
> Yea that would be more flexible. In terms of the API, I think we'd want something like timeRemaining (or maybe currentTime), as well as something like totalTime (or initialTime, or targetTime). Thoughts on naming those props? Should we use "seconds" instead of "time" to be more explicit?

Accepting a float between 0-1 is probably the 'cleanest' api, but maybe consider aligning with the existing progress-bar pattern. I guess this does not support rendering the time-remaining. Number of seconds remaining seems appropriate

File: src/button-timed/types.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  BaseButtonTimed?: Override;
+  TimerContainer?: Override;
+};
+
+export type ButtonTimedProps = Omit<
+  ButtonProps,
+  'kind' | 'shape' | 'size' | 'onClick' | 'overrides'
+> & {
+  time: number;
+  onClick: (a?: React.SyntheticEvent<HTMLButtonElement>) => unknown;
```
[the remark this answers is not in the document]
Comment (lhbrennan) [reply src/button-timed/types.ts:20]: Design wants to make sure the same thing happens when the button is clicked vs the time runs out. `onClick` seems like it's what devs would expect this to be called, but maybe there's something better that captures both the idea of clicking and of time running out?

File: src/button-timed/button-timed.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    BaseButtonTimed: BaseButtonTimedOverride,
+    TimerContainer: TimerContainerOverride,
+    ...buttonOverrides
+  } = overrides;
+
+  const [secondsRemaining, setSecondsRemaining] = React.useState<number>(time);
+
+  React.useEffect(() => {
+    let interval = null;
+    interval = setInterval(() => {
```
[the remark this answers is not in the document]
Comment (lhbrennan) [reply src/button-timed/button-timed.tsx]: Updated, thanks.

**Review by chasestarr — APPROVED** [review]
lgtm after some more integration tests

**Pull request merged**